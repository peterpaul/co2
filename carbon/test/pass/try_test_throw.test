["stdio.h"] {
	int printf (char *format, ...);
}

class Ex {
	char * msg;

	Ex (char * msg) {
		self.msg = msg;
	}
}

int fun (int in) {
	printf ("in: %d\n", in);
	try {
		if (in %2 == 0)
			throw new Ex ("Should display in fun");
	} catch (Ex e) {
		printf ("fun caught Ex: %s\n", e.msg);
		return in + 1;
	} finally {
		throw new Ex ("should display in main");
	}
	printf ("FATAL: this should never display!\n");
	return -1;
}

int loop (int val) {
	try {
		int out = fun (val);
		printf ("out: %d\n", out);
		throw new Ex ("Should display in main");
	} catch (Ex e) {
		printf ("main caught Ex: %s\n", e.msg);
	} finally {
		return 0;
	}
}

int main () {
	printf ("---\n");
	loop (0);
	printf ("---\n");
	loop (1);
	printf ("---\n");
	return 0;
}
