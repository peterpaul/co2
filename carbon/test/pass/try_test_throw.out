---
in: 0
fun caught Ex: Should display in fun
main caught Ex: should display in main
---
in: 1
main caught Ex: should display in main
---
